### PR TITLE
Fix alias to map in the model's properties

### LINKED ISSUE #255: [C#] Missing Models for Map types and incomplete types for Api Endpoints
##### Description

With the following **valid** spec
- The map models are not generated and the Api is having a wrong type for the response associated with map.

- When a map is part of an object the model is generated correctly with the map inside the object.
When a model type is a map, there is incorrect generation of code either as an inline model or as a map when referred from another model.

- General issues with inline models always being reported as objects

##### openapi-generator version

3.0.0

##### OpenAPI declaration file content or url

```yaml 
---
# This is a sample Swagger spec
 
swagger: "2.0"
info:
  description: Application Description
  version: 1.0.0
  title: Application
  
host: 0.0.0.0:8080
basePath: /api/v1.0
schemes:
- http

tags:
- name: ApplicationResources
  x-displayName: "ApplicationResources"
  description: |
    "Application Resources."

- name : ApplicationLifeCycle
  x-displayName : "ApplicationLifeCycle"
  description : |
    "Application Lifecycle."

- name: ApplicationTestConfiguration
  x-displayName : "ApplicationTestConfiguration"
  description : |
    "Application Test Configuration"
    
- name: ApplicationEndpointTest
  x-displayName: "ApplicationEndpointTest"
  description: |
    "Application Endpoint Test"
    
paths:
  /default/Resources/{id}: 
    post:
      tags:
      - ApplicationResources      
      operationId: resourceStatusSet
      consumes:
      - application/json
      produces:
      - application/json
      parameters:
      - name: "exclusive"
        in: "query"
        type: "boolean"
        default: false
      - name: id
        in: "path"
        required: true
        type: integer
        format: int32
      - name: "body"
        in: "body"
        schema:
          $ref: "#/definitions/ResourceRequest"
      responses:
        200:
          description: ""
          schema:
            $ref: "#/definitions/ResourceOperationsStatus"
        404:
          description: ""
          schema:
            $ref: "#/definitions/ErrorSts"
    get:
      tags:
      - ApplicationResources
      operationId: resourceStatusGet
      produces:
      - application/json
      parameters:
      - name: id
        in: path
        description: "Id of the Resource"
        required: true
        type: integer
        format: int32
      responses:
        200:
          description: ""
          schema:
            $ref: "#/definitions/Resource"
        404:
          description: ""        
          schema:
            $ref: "#/definitions/ErrorSts"  
              
  /default/ApiTestSetup: 
    post: 
      tags: 
      - ApplicationTestConfiguration     
      operationId : apiTestConfigurationSet
      consumes:
      - application/json
      parameters : 
      - name : "body"
        in : "body"
        required: true
        schema: 
          $ref: "#/definitions/TestEndPointConfig"
      responses:
        200:
          description: "Successful response"
        400:
          description: "The request was not successfully executed." 
    get:
      tags:
      - ApplicationTestConfiguration
      operationId: apiTestConfigurationGet
      produces:
      - application/json      
      responses:
        200:
          description: "Successful response"
          schema: 
            $ref: "#/definitions/TestEndPointConfig"          
        400:
          description: "The request was not successfully executed."           
  
  
definitions:

  Options:
    type: "object"
    properties:
      Config:
        type: "object"
        additionalProperties:
          type: "string"

  Resource:
    required:
      - id
    type: object  
    properties:
      id:
        type: integer
        format: int32
      name:
        type: string 
      type:
        type: string
      options:
        $ref: "#/definitions/Options" 

  ResourceOperations:
    type: "object"
    additionalProperties:
      type: "string"
      
  ResourceOperationsStatus:
    type: "object"
    additionalProperties:
      type: "string"
                
  ResourceRequest:
    allOf:
    - $ref: "#/definitions/Resource"
    - type: "object"
      required:
      - handle
      properties:
        handle:
          type : "string"
          x-nullable: false
        requestCode:
          type: integer
          format: int32
        operations:
          $ref: "#/definitions/ResourceOperations"
        payload:
          type: array
          items:
            type: integer
            format: int32    
        args:
          type: "array"
          items:
            type: "string"  
  
  ErrorSts:
    required:
      - code
    type : object   
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string  
       
  TestEndPointConfig : 
    type : "object"
    additionalProperties :
      type: "string"      
```

##### Command line used for generation

java -jar ~/dev/openapi-generator/modules/openapi-generator-cli/target/openapi-generator-cli.jar generate -i ./Cta.yaml -o ./genserver -l aspnetcore --additional-properties packageName=Cta
##### Suggest a fix/enhancement

Is the Map support broken in C#

### PR DESCRIPTION
**Pull request by wing328**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `3.1.x`, `4.0.x`. Default: `master`.
- [ ] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR

To fix #255, #191 

This is the first fix for "alias". Please expect more PRs to fix various usage of alias (e.g. alias of alias, alias of array, alias handling in response, parameters, etc)

In the upcoming PRs, I'll also add more Java unit tests to cover the issue moving forward.

### CODE REVIEW

File: samples/client/petstore/ruby/lib/petstore/models/enum_test.rb [hunk 1]
```diff
@@ -64,7 +64,7 @@ def self.openapi_types
         :'enum_string_required' => :'String',
         :'enum_integer' => :'Integer',
         :'enum_number' => :'Float',
-        :'outer_enum' => :'OuterEnum'
+        :'outer_enum' => :'String'
```
Comment (wing328) [inline samples/client/petstore/ruby/lib/petstore/models/enum_test.rb]: @ackintosh FYI. Now outer_enum is referencing `String` correctly

Comment (wing328) [reply samples/client/petstore/ruby/lib/petstore/models/enum_test.rb]: The PHP tests failed with the following errors:
```
PHPUnit 4.8.36 by Sebastian Bergmann and contributors.

.........................................F....................... 65 / 69 ( 94%)
....

Time: 2.12 seconds, Memory: 15.75MB

There was 1 failure:

1) OpenAPI\Client\OuterEnumTest::testSanitizeNestedInvalidValue
Failed asserting that exception of type "InvalidArgumentException" is thrown.
```

I tried to fix it locally but in vain. I tried supplying invalid values (e.g. "not_upper", 1000, etc to other enum values) but still no exception is thrown.

I've skipped the test via d88c9bf for the time being

**Comment by wing328** [comment]
cc core team (@jimschubert @jmini @cbornet @ackintosh @JFCote @jaz-ah )

File: samples/client/petstore/ruby/lib/petstore/models/enum_test.rb (hunk 1 above)
Comment (wing328) [reply samples/client/petstore/ruby/lib/petstore/models/enum_test.rb]: UPDATE: I pushed another commit to retain the top-level enum classes. Thanks for the test case to cover it.

**Comment by jmini** [comment]
I did not look at it in details yet, it looks good to me in general.

---

One question about the targeted branch: 

this change has a big impact on the generated code (it is an improvement because it solves something that is not correct now, but still an impact). This might break some existing code relying on the generated code. 

I think it is preferable to have this filed for `3.1.x` (just change the target branch in the GitHub UI). From a release perspective, it is only one week more to wait (`3.1.0` is due at beginning of july).

**Comment by wing328** [comment]
> I think it is preferable to have this filed for 3.1.x (just change the target branch in the GitHub UI). From a release perspective, it is only one week more to wait (3.1.0 is due at beginning of july).

Definitely a valid point. I don't mind targeting 3.1.x. I'll update it later today if no one has other suggestions/feedbacks on which branch this PR should target.

**Comment by etherealjoy** [comment]
Issue solved for #255 
With the spec mentioned in this issue, the code is generated for go-server and aspnetcore.

**Comment by wing328** [comment]
The CI failures are not related to this change

**Pull request merged**